HDDS-6627. Intermittent failure in TestGetCommittedBlockLengthAndPutKey#tesGetCommittedBlockLength

devmadhuu opened the pull request:
## What changes were proposed in this pull request?
This PR is to fix the flakiness in `TestGetCommittedBlockLengthAndPutKey#tesGetCommittedBlockLength`. Here the test case doing writeChunks call and then `putBlock` which should save the blockId in BlockData table of container DB, but this persistence of blockdata seems taking time by fraction of second due to which immediate `GetCommittedBlockLength` call by test client sometime fail and throws exception below:
```
org.apache.hadoop.hdds.scm.container.common.helpers.StorageContainerException: Unable to find the block. BlockID : conID: 1 locID: 111486602482745345 bcsId: 0
	at org.apache.hadoop.ozone.container.keyvalue.impl.BlockManagerImpl.getBlockByID(BlockManagerImpl.java:359)
	at org.apache.hadoop.ozone.container.keyvalue.impl.BlockManagerImpl.getCommittedBlockLength(BlockManagerImpl.java:284)
	at org.apache.hadoop.ozone.container.keyvalue.KeyValueHandler.handleGetCommittedBlockLength(KeyValueHandler.java:622)
```
This PR changes tries to fix the flakiness by adding waitFor `GetCommittedBlockLength` call till the response provides blockId successfully.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-6627

## How was this patch tested?
This patch was tested using repeated CI run with 500 iterations. Here is the green CI [link](https://github.com/devmadhuu/ozone/actions/runs/7020463753).

devmadhuu commented:
@sumitagrawl Pls review

nandakumar131 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -110,12 +115,21 @@ public void tesGetCommittedBlockLength() throws Exception {
+            .getCommittedBlockLength(client, blockID, null);
+      } catch (IOException e) {
+        success = false;
+        LOG.error("Ignore the exception till wait: {}", e.getMessage());
```
There is no need to log `error` here. It can be changed to `debug` if this will help with debugging in case of failure.

devmadhuu replied:
Thank @nandakumar131 for review. Changes done.

nandakumar131 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -110,12 +115,21 @@ public void tesGetCommittedBlockLength() throws Exception {
-    response = ContainerProtocolCalls
-        .getCommittedBlockLength(client, blockID, null);
+    GenericTestUtils.waitFor(() -> {
+      boolean success = true;
```
There is no need for a `boolean` variable in this scenario. We can directly return `true` or `false` when needed.

devmadhuu replied:
Thank @nandakumar131 for review. Changes done.

ayushtkn commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -110,12 +115,21 @@ public void tesGetCommittedBlockLength() throws Exception {
     // make sure the block ids in the request and response are same.
     Assertions.assertTrue(
-        BlockID.getFromProtobuf(response.getBlockID()).equals(blockID));
-    Assertions.assertTrue(response.getBlockLength() == data.length);
+        BlockID.getFromProtobuf(response[0].getBlockID()).equals(blockID));
+    Assertions.assertTrue(response[0].getBlockLength() == data.length);
```
This should be ``assertEquals`` I believe rather than doing equals inside ``assertTrue``

```
    Assertions.assertEquals(BlockID.getFromProtobuf(response[0].getBlockID()), blockID);
    Assertions.assertEquals(response[0].getBlockLength(), data.length);
```

devmadhuu replied:
Thanks @ayushtkn for review, I have changed it to assertEquals.

ayushtkn commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -110,12 +115,21 @@ public void tesGetCommittedBlockLength() throws Exception {
+        LOG.error("Ignore the exception till wait: {}", e.getMessage());
+      }
+      return success;
+    }, 1000, 5000);
```
this would be 5 tries I believe, did you try how much time does it take, if it is fast you can explore reducing 1000 to a lower number

devmadhuu replied:
What i observed was not very long time, it appears to be in millisec. so good idea to change to lower down the wait check value. I changed now to 500ms.

ayushtkn commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -110,12 +115,20 @@ public void tesGetCommittedBlockLength() throws Exception {
-    response = ContainerProtocolCalls
-        .getCommittedBlockLength(client, blockID, null);
+    GenericTestUtils.waitFor(() -> {
+      try {
+        response[0] = ContainerProtocolCalls
+            .getCommittedBlockLength(client, blockID, null);
+      } catch (IOException e) {
+        LOG.debug("Ignore the exception till wait: {}", e.getMessage());
+        return false;
+      }
+      return true;
+    }, 500, 5000);
     // make sure the block ids in the request and response are same.
-    Assertions.assertTrue(
-        BlockID.getFromProtobuf(response.getBlockID()).equals(blockID));
-    Assertions.assertTrue(response.getBlockLength() == data.length);
+    Assertions.assertEquals(BlockID.getFromProtobuf(response[0].getBlockID()),
+        blockID);
+    Assertions.assertEquals(response[0].getBlockLength(), data.length);
```
I think this is flipped, the expected goes first, just double check once

ayushtkn reviewed (approved):
Check on the order of assertions, 
rest LGTM

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -42,21 +42,26 @@
  * Test Container calls.
  */
-@Timeout(300)
+@Timeout(5000)
```
If the flakiness is fixed by waiting at most 5 seconds in the test case, this increase of timeout seems unnnecessary.  5000 seconds is more than an hour.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java:
```diff
@@ -85,9 +90,9 @@ public static void shutdown() throws InterruptedException {
-  @Flaky("HDDS-6627")
   public void tesGetCommittedBlockLength() throws Exception {
-    ContainerProtos.GetCommittedBlockLengthResponseProto response;
+    final ContainerProtos.GetCommittedBlockLengthResponseProto[] response =
+        new ContainerProtos.GetCommittedBlockLengthResponseProto[1];
```
Let's use `AtomicReference` instead of a single-element array.

devmadhuu commented:
> HDDS-9705

Thanks @adoroszlai , I am analyzing both again.

devmadhuu commented:
@ayushtkn @adoroszlai , I have fixed it now. Here is the green CI [link](https://github.com/devmadhuu/ozone/actions/runs/7043089744). 

There were actually 2 issues:

1. Timeout issue [HDDS-9705](https://issues.apache.org/jira/browse/HDDS-9705) where sendCommand of writeChunk fails.
2. And another one when chunk was failed to write, GetCommittedBlockLength command fails to get block due to unsuccessful chunk write in `tesGetCommittedBlockLength`.

I just realized that line L173 change in `TestGetCommittedBlockLengthAndPutKey` which I did in my local code to fix the flakyness, but was somehow reverted and was not pushed in this PR, now I have fixed that. Above green CI link shows no falkiness now.

adoroszlai reviewed (approved):
Thanks @devmadhuu for the fix.

adoroszlai reviewed (changes requested):
Sorry, I just noticed: can you please fix `nextInt(0, 1024)` in line 108, too?

https://github.com/apache/ozone/blob/8ca699d11bede96e619990e73e428a791bc2ad2e/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java#L107-L108

devmadhuu commented:
> Sorry, I just noticed: can you please fix `nextInt(0, 1024)` in line 108, too?
> 
> https://github.com/apache/ozone/blob/8ca699d11bede96e619990e73e428a791bc2ad2e/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestGetCommittedBlockLengthAndPutKey.java#L107-L108

Thanks @adoroszlai for pointing out. I just fixed it at line 108 as well.
